### PR TITLE
Add option to clean div B in ICs and enable MHD DiskGalaxy

### LINKED ISSUE #1250: compute divergence-free B field numerically
**Describe the proposal**
Some magnetic field configurations do not have closed-form solutions for their vector potential. In particular, this is the case for a toroidal magnetic field [1].

For this case, we'd like to first compute B on faces using a closed-form expression (knowing that this B will NOT discretely satisfy div B = 0), then compute the vector potential numerically, and then finally, compute the divergence-free B using finite differences.

We can do this AMReX's curl-curl linear solver [2,3] by writing the vector potential equation in the form:
$\nabla \times (\nabla \times \vec{A}) = \vec{J}$
where
$\vec{J} = \nabla \times \vec{B}$.

Note that $A$ is edge-centered. $B$ is face-centered, and $J$ is node-centered.

With Dirichlet boundary conditions, this is well-posed. Since we don't care about the gauge, we can just set the boundary conditions to zero.

**Describe alternatives you've considered**
An alternative would be to use the MAC projector to compute an approximate divergence-free B directly, but div B is only satisfied in that case to the tolerance of the solver, rather than to machine precision. This will probably cause the simulation to blow up.

**Additional context**
This is needed for the DiskGalaxy problem.

[1] https://iopscience.iop.org/article/10.1088/0305-4470/26/3/030
[2] https://amrex-codes.github.io/amrex/doxygen/classamrex_1_1MLCurlCurl.html#details
[3] https://github.com/AMReX-Codes/amrex/pull/3682

### PR DESCRIPTION
**Pull request by BenWibking**
### Description
Uses the `MacProjector` from AMReX-Hydro to project out the divergence from the initial face-centered B field to machine precision. (This uses `amrex::MLMG` internally, so it may not be bitwise reproducible for multiple AMR levels, since it uses floating-point atomics to do coarse-fine reflux on GPU.)

This is enabled at runtime with:
```
mhd.project_initial_b_field = true
mhd.update_initial_b_energy = true
```

The latter option resets the total energy to agree with magnetic energy of the cleaned B-field.

Since this feature is needed to initialize a toroidal field, MHD is now enabled for the `DiskGalaxy` problem in this PR.

### Related issues
Closes https://github.com/quokka-astro/quokka/issues/1250.

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [x] I have added a description (see above).
- [x] I have added a link to any related issues (if applicable; see above).
- [x] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [ ] I have added tests for any new physics that this PR adds to the code.
- [ ] *(For quokka-astro org members)* I have manually triggered the GPU tests with the magic comment `/azp run`.

### CODE REVIEW

**Comment by BenWibking** [comment]
Need to fix Dirichlet BCs for cleaning solve:
```
• Review comment:

  - [P0] Preserve Dirichlet B boundary values during projection — src/QuokkaSimulation.hpp:1238-1280
    When BCs_fc_ contains ext_dir entries we map them to LinOpBCType::Dirichlet, but we then call macproj.setLevelBC(lev, nullptr).
  In Hydro::MacProjector a null pointer means the Dirichlet value is zero, so the projection step will overwrite any non-zero boundary
  data. For example, the Brio–Wu shock tube sets Bx to ±0.75 at the left/right faces via ext_dir; after the projection this code
  will force those faces to zero, corrupting the initial state. Please either supply the actual boundary data to setLevelBC or avoid
  projecting when Dirichlet faces are present.
```

**Comment by BenWibking** [comment]
clang-tidy doesn't like case statements. Probably should remove it:
```
/github/workspace/src/QuokkaSimulation.hpp:1254:4: warning: switch has 2 consecutive identical branches [bugprone-branch-clone]
 1254 |                         case amrex::BCType::ext_dir:
      |                         ^
/github/workspace/src/QuokkaSimulation.hpp:1259:41: note: last of these clones ends here
 1259 |                                 return amrex::LinOpBCType::Dirichlet;
      |                                                                     ^
```

**Comment by BenWibking** [comment]
@codex Review this PR.

**Comment by BenWibking** [comment]
```
• Review comment:

  - [P1] Fix magnetic BC parity — src/problems/DiskGalaxy/galaxy.cpp:443-447
    For a reflecting boundary, the face-centered magnetic component normal to the boundary must be reflected evenly (no sign flip) while the
  tangential components should be reflected oddly. The new loop is doing the opposite: when component_dir == idim we set reflect_odd, which
  forces the normal field component to change sign and therefore vanish at the boundary, breaking the intended toroidal field and introducing a
  large divergence error right at startup. Please swap the parity assignment so that the normal component uses reflect_even and the tangential
  components use reflect_odd.
```

**Comment by BenWibking** [comment]
I think there was a bad commit:
```

Coarse STEP 1 at t = 0 (0%) starts ...
...[level 0] estimated hydro timestep: 3.702340e+09
...[level 0] 	hydro timestep limited at cell [59, 65, 66] with signal speed = 1.562699e+12
...[level 0] 	cell density = 9.963234e-55, |v| = 0.000000e+00, cs = 9.364488e+11
...[level 0] estimated particle timestep: 1.797693e+308
...[level 0] 	max particle velocity: 1.843763e+07
...[level 0] 	particle timestep limited at position [-9.332715e+23, 2.029194e+23, 4.685295e+22]
...[level 0] timestep limited by HYDRO
...[level 1] estimated hydro timestep: 1.292464e+09
...[level 1] 	hydro timestep limited at cell [119, 130, 123] with signal speed = 2.238223e+12
...[level 1] 	cell density = 9.963234e-55, |v| = 0.000000e+00, cs = 1.349693e+12
...[level 1] estimated particle timestep: 1.797693e+308
...[level 1] 	max particle velocity: 2.523782e+07
...[level 1] 	particle timestep limited at position [4.636320e+23, 2.509672e+21, 5.905351e+22]
...[level 1] timestep limited by HYDRO
...[level 2] estimated hydro timestep: 5.164574e+08
...[level 2] 	hydro timestep limited at cell [261, 244, 246] with signal speed = 2.800640e+12
...[level 2] 	cell density = 9.963234e-55, |v| = 0.000000e+00, cs = 1.674944e+12
...[level 2] estimated particle timestep: 1.797693e+308
...[level 2] 	max particle velocity: 3.179644e+07
...[level 2] 	particle timestep limited at position [2.381067e+23, -6.024940e+21, -5.316292e+22]
...[level 2] timestep limited by HYDRO
...[level 3] estimated hydro timestep: 1.977710e+08
...[level 3] 	hydro timestep limited at cell [521, 489, 494] with signal speed = 3.656784e+12
...[level 3] 	cell density = 9.963234e-55, |v| = 0.000000e+00, cs = 2.186411e+12
...[level 3] estimated particle timestep: 1.797693e+308
...[level 3] 	max particle velocity: 3.872657e+07
...[level 3] 	particle timestep limited at position [-1.196662e+23, 9.956788e+21, -1.355480e+22]
...[level 3] timestep limited by HYDRO
...[level 4] estimated hydro timestep: 9.335840e+07
...[level 4] 	hydro timestep limited at cell [1045, 970, 990] with signal speed = 3.873276e+12
...[level 4] 	cell density = 9.963234e-55, |v| = 0.000000e+00, cs = 2.315041e+12
...[level 4] estimated particle timestep: 1.797693e+308
...[level 4] 	max particle velocity: 4.448629e+07
...[level 4] 	particle timestep limited at position [4.713705e+20, -1.682594e+22, -6.329401e+22]
...[level 4] timestep limited by HYDRO
...[level 5] estimated hydro timestep: 9.714088e+08
...[level 5] 	hydro timestep limited at cell [1952, 2143, 1984] with signal speed = 1.861229e+11
...[level 5] 	cell density = 2.511310e-52, |v| = 1.987247e+07, cs = 1.112178e+11
...[level 5] estimated particle timestep: 8.935548e+12
...[level 5] 	max particle velocity: 5.058487e+07
...[level 5] 	particle timestep limited at position [7.595951e+21, -1.161089e+21, 3.323199e+22]
...[level 5] timestep limited by HYDRO
...[level 6] estimated hydro timestep: 3.659968e+12
...[level 6] 	hydro timestep limited at cell [4137, 4044, 4095] with signal speed = 2.469986e+07
...[level 6] 	cell density = 3.133969e-25, |v| = 2.317983e+07, cs = 1.519836e+06
...[level 6] estimated particle timestep: 4.237559e+12
...[level 6] 	max particle velocity: 5.333301e+07
...[level 6] 	particle timestep limited at position [8.715361e+21, -1.438639e+21, 2.317156e+22]
...[level 6] timestep limited by HYDRO
...[level 7] estimated hydro timestep: 1.829964e+12
...[level 7] 	hydro timestep limited at cell [8290, 8103, 8191] with signal speed = 2.470013e+07
...[level 7] 	cell density = 3.407766e-25, |v| = 2.318007e+07, cs = 1.519849e+06
...[level 7] estimated particle timestep: 2.031062e+12
...[level 7] 	max particle velocity: 5.563637e+07
...[level 7] 	particle timestep limited at position [6.949743e+21, 5.883385e+21, -1.525289e+22]
...[level 7] timestep limited by HYDRO
...[level 8] estimated hydro timestep: 9.149789e+11
...[level 8] 	hydro timestep limited at cell [16578, 16203, 16384] with signal speed = 2.470022e+07
...[level 8] 	cell density = 3.556998e-25, |v| = 2.318013e+07, cs = 1.519855e+06
...[level 8] estimated particle timestep: 8.523480e+11
...[level 8] 	max particle velocity: 6.628800e+07
...[level 8] 	particle timestep limited at position [6.740506e+19, -1.875805e+20, 2.384845e+19]
...[level 8] timestep limited by PARTICLES
...coarse timestep set by level 4
[Level 0 step 1] ADVANCE with time = 0.0000000000e+00 dt = 9.3358396996e+07
	cooling substeps (per cell): avg 1, max 1
[FOFC-2] flux correcting 930 cells on level 0
Coordinates: (-7.232056846e+22, -2.169617054e+23, -1.012487958e+23):  At cell (61,56,60) in Box ((0,0,0) (63,63,63) (0,0,0)): -5.9622926401207054e-54, 4.052100973369509e-48, 4.2087055359978279e-48, 1.0974103343140658e-47, 5.9415553486070574e-33, 1.1426766243111362e-38
[FOFC-2] Flux correction failed:
Coordinates: (-1.591052506e+23, -1.591052506e+23, -1.301770232e+23):  At cell (58,58,59) in Box ((0,0,0) (63,63,63) (0,0,0)): -1.4950419114400508e-54, 6.8131894187627817e-49, 6.8131896907750981e-49, 2.2360650822128072e-35, 1.6490518033590748e-33, -8.5816781301790024e-39
[FOFC-2] failed for 256 cells on level 0
	>> Re-trying hydro advance at level 0 with reduced timestep (remaining dt = 93358397, total substeps = 2, completed substeps = 0, dt_new = 46679198.5)
	cooling substeps (per cell): avg 1, max 1
[FOFC-2] flux correcting 852 cells on level 0
Coordinates: (-7.232056846e+22, -1.88033478e+23, -1.301770232e+23):  At cell (61,57,59) in Box ((0,0,0) (63,63,63) (0,0,0)): -5.2320709260536829e-54, 1.0712228915624138e-48, 4.191632282734093e-49, 3.1756113757799022e-48, 4.4964163343256548e-33, -2.5156185613762034e-39
[FOFC-2] Flux correction failed:
Coordinates: (-1.591052506e+23, -1.591052506e+23, -1.301770232e+23):  At cell (58,58,59) in Box ((0,0,0) (63,63,63) (0,0,0)): -2.4935923571925916e-55, 3.4069802131019227e-49, 3.4069803491109265e-49, 1.1180325411064036e-35, 1.6487997814200227e-33, -4.1872904050895012e-39
[FOFC-2] failed for 168 cells on level 0
	>> Re-trying hydro advance at level 0 with reduced timestep (remaining dt = 93358397, total substeps = 4, completed substeps = 0, dt_new = 23339599.25)
	cooling substeps (per cell): avg 1, max 1
[FOFC-2] flux correcting 704 cells on level 0
Coordinates: (-2.169617054e+23, -7.232056846e+22, -1.012487958e+23):  At cell (56,61,60) in Box ((0,0,0) (63,63,63) (0,0,0)): -7.4333057958006137e-55, 1.0533044938553055e-48, 1.0131306959114102e-48, 2.7451427439714739e-48, 5.9410782940431587e-33, 3.0120145878662375e-39
[FOFC-2] Flux correction failed:
Coordinates: (-1.012487958e+23, -1.88033478e+23, -1.012487958e+23):  At cell (60,57,60) in Box ((0,0,0) (63,63,63) (0,0,0)): -4.7488381487547002e-54, 5.5901627055304089e-36, 5.5901627055293485e-36, 1.9044479080077177e-48, 9.572187686919488e-33, -3.5348515549448864e-38
[FOFC-2] failed for 80 cells on level 0
	>> Re-trying hydro advance at level 0 with reduced timestep (remaining dt = 93358397, total substeps = 8, completed substeps = 0, dt_new = 11669799.62)
	cooling substeps (per cell): avg 1, max 1
[FOFC-2] flux correcting 604 cells on level 0
Coordinates: (-7.232056846e+22, -1.591052506e+23, -1.591052506e+23):  At cell (61,58,58) in Box ((0,0,0) (63,63,63) (0,0,0)): -1.1782467527185581e-54, 9.5976225022373875e-50, 9.0618188196161121e-50, 4.7888073868808974e-49, 1.8550464019462841e-33, -2.0103951530716505e-39
[FOFC-2] Flux correction failed:
Coordinates: (-1.446411369e+22, -1.591052506e+23, -1.012487958e+23):  At cell (63,58,60) in Box ((0,0,0) (63,63,63) (0,0,0)): -1.3618921271617603e-53, -2.7950813527615114e-36, 2.7950813527636623e-36, 1.316858610188588e-47, 4.0093068688268352e-32, 7.2646361017392485e-38
[FOFC-2] failed for 56 cells on level 0
	>> Re-trying hydro advance at level 0 with reduced timestep (remaining dt = 93358397, total substeps = 16, completed substeps = 0, dt_new = 5834899.812)
	cooling substeps (per cell): avg 1, max 1
[FOFC-2] flux correcting 504 cells on level 0
Coordinates: (-1.591052506e+23, -7.232056846e+22, -1.591052506e+23):  At cell (58,61,58) in Box ((0,0,0) (63,63,63) (0,0,0)): -9.0961656378602444e-56, 4.530890761327273e-50, 4.7988050663533802e-50, 2.3943998120109239e-49, 1.8550493660481485e-33, -9.0164891877281858e-40
[FOFC-2] Flux correction failed:
Coordinates: (-1.012487958e+23, -1.591052506e+23, -1.301770232e+23):  At cell (60,58,59) in Box ((0,0,0) (63,63,63) (0,0,0)): -1.9297808590125626e-55, 1.3975406763828399e-36, 1.3722301712462018e-49, 3.7595832330367229e-49, 5.3687293612742438e-33, -1.7087568298827539e-39
[FOFC-2] failed for 48 cells on level 0
	>> Re-trying hydro advance at level 0 with reduced timestep (remaining dt = 93358397, total substeps = 32, completed substeps = 0, dt_new = 2917449.906)
	cooling substeps (per cell): avg 1, max 1
[FOFC-2] flux correcting 464 cells on level 0
Coordinates: (-4.339234108e+22, -1.446411369e+22, -7.232056846e+22):  At cell (62,63,61) in Box ((0,0,0) (63,63,63) (0,0,0)): -9.4490465944074965e-46, 1.1694320520308521e-43, -3.4281547002880321e-43, 1.4360011932925944e-40, 2.3723531364641673e-31, 9.4802231102830722e-34
[FOFC-2] Flux correction failed:
Coordinates: (-7.232056846e+22, -7.232056846e+22, -1.012487958e+23):  At cell (61,61,60) in Box ((0,0,0) (63,63,63) (0,0,0)): -6.6565049921794385e-55, -8.3560649410931236e-49, -8.6942933109215604e-49, 6.9877033815855223e-37, 8.5942524883999916e-32, 6.7159286278244015e-35
[FOFC-2] failed for 40 cells on level 0
	>> Re-trying hydro advance at level 0 with reduced timestep (remaining dt = 93358397, total substeps = 64, completed substeps = 0, dt_new = 1458724.953)
	cooling substeps (per cell): avg 1, max 1
[FOFC-2] flux correcting 368 cells on level 0
Coordinates: (-7.232056846e+22, -1.012487958e+23, -1.012487958e+23):  At cell (61,60,60) in Box ((0,0,0) (63,63,63) (0,0,0)): -1.1302734210333179e-53, 3.0733812142129095e-49, -6.3942575181898392e-49, 4.5987155319602508e-48, 1.052293563335578e-31, 4.1869939460626035e-37
[FOFC-2] Flux correction failed:
Coordinates: (-4.339234108e+22, -1.88033478e+23, -1.012487958e+23):  At cell (62,57,60) in Box ((0,0,0) (63,63,63) (0,0,0)): -1.2317704481373049e-55, -1.1627860250335337e-49, 3.4938516909527245e-37, 3.347795957990181e-49, 1.8551564474848224e-32, -5.4213890059401554e-39
[FOFC-2] failed for 64 cells on level 0

QUOKKA FATAL ERROR
Hydro update exceeded max_retries on level 0. Cannot continue, crashing...

```

**Comment by BenWibking** [comment]
This PR has the same Sonarcloud issue as https://github.com/quokka-astro/quokka/pull/1481#issuecomment-3620768882.

**Comment by markkrumholz** [comment]
It looks like there are changes in here that have been imported from #1481. I'll wait until that is merged in before finishing a review of this, so that we can cleanly separate the changes that are specific to this PR.

**Comment by BenWibking** [comment]
@chongchonghe @markkrumholz This PR should be ready for review.

File: src/problems/DiskGalaxy/testDiskGalaxy.cpp [hunk 1]
```diff
@@ -267,11 +285,64 @@ template <> void QuokkaSimulation<AgoraGalaxy>::setInitialConditionsOnGrid(quokk
...
+
+	// disc parameters
+	double disk_Rscale_kpc = NAN; // disk scale length
+	double disk_zscale_kpc = NAN; // disk scale height
+	pp.query("disk_Rscale_kpc", disk_Rscale_kpc);
+	pp.query("disk_zscale_kpc", disk_zscale_kpc);
+	AMREX_ALWAYS_ASSERT(!std::isnan(disk_Rscale_kpc));
+	AMREX_ALWAYS_ASSERT(!std::isnan(disk_zscale_kpc));
+	const double R_d = disk_Rscale_kpc * (1.0e3 * C::parsec);
+	const double z_d = disk_zscale_kpc * (1.0e3 * C::parsec);
```
[the remark this answers is not in the document]
Comment (markkrumholz) [reply src/problems/DiskGalaxy/testDiskGalaxy.cpp:307]: I don't think this is particularly important, since the cost of parsing again is negligible. I will dismiss this suggestion.

File: src/QuokkaSimulation.hpp [hunk 2]
```diff
@@ -1084,6 +1103,314 @@ template <typename problem_t> void QuokkaSimulation<problem_t>::applyPoissonGrav
...
+			amrex::Array<amrex::MultiFab const *, AMREX_SPACEDIM> bfield_ptrs;
+			for (int dir = 0; dir < AMREX_SPACEDIM; ++dir) {
+				bfield_ptrs[dir] = bfield_storage[lev][dir].get();
+			}
+			amrex::computeDivergence(divB, bfield_ptrs, geom_levels[lev]);
+
+			auto const &dx = geom_levels[lev].CellSizeArray();
+			amrex::Real const dx_min = std::min({AMREX_D_DECL(dx[0], dx[1], dx[2])});
+			amrex::Real const scale_b = std::max(max_bmag_global, small_b);
+
```
[the remark this answers is not in the document]
Comment (markkrumholz) [reply src/QuokkaSimulation.hpp:1477]: This does seem like a good suggestion to implement. This will be a practical issue in dynamo simulations, for example, where the initial conditions typically do begin with a very, very small field, E_mag / E_k ~ 10^-20.

**Review by markkrumholz — CHANGES_REQUESTED** [review]
I do think one of the AI suggestions is worth implementing: handling the case when the initial field is so small that the target relative tolerance in div B cannot be reached due to machine precision. Other than that, I'm ready to approve this PR.

File: src/QuokkaSimulation.hpp [hunk 3]
```diff
@@ -514,6 +527,13 @@ template <typename problem_t> void QuokkaSimulation<problem_t>::readParmParse()
 		hpp.query("emf_averaging_scheme", emfAveragingScheme_);
 	}
 
+	// set Quokka runtime parameters
```
Comment (chongchonghe) [inline src/QuokkaSimulation.hpp]: Why aren't these parameters in "mhd" group? They seem natural fit to mhd group.

**Review by chongchonghe — COMMENTED** [review]
Can you update the PR title and description to describe what you have done comprehensively? This PR also turns on MHD in DiskGalaxy and this is worth noticing. 

Just a note: The new `projectFaceCenteredMagneticField` function uses `amrex::MLMG`, so we want to keep an eye on the potential non-reproducibility introduced.

File: src/QuokkaSimulation.hpp (hunk 3 above)
Comment (BenWibking) [reply src/QuokkaSimulation.hpp]: That's a good point. I'll change it.

Comment (BenWibking) [reply src/QuokkaSimulation.hpp]: Fixed.

File: src/QuokkaSimulation.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenWibking) [reply src/QuokkaSimulation.hpp:1477]: I'm not sure whether this is a real issue. I think it should be able to clean the divergence even when the absolute value of the field is very small, as long as it is significantly above the double precision floor ~ 1e-300.

If it is the case, then I think that means the units of the problem need to be rescaled.

**Comment by BenWibking** [comment]
> I do think one of the AI suggestions is worth implementing: handling the case when the initial field is so small that the target relative tolerance in div B cannot be reached due to machine precision. Other than that, I'm ready to approve this PR.

@markkrumholz I think the AI is wrong about this issue. I don't think the divergence cleaning is limiting by the overall scaling of the B field itself. See my inline comment.

**Review by markkrumholz — APPROVED** [review]

**Comment by markkrumholz** [comment]
This looks like a transient failure. Attempting again.
